src/Concordium/Client/Runner.hs

```diff
@@ -3438,95 +3572,253 @@ processIdentityShowCmd :: IdentityShowCmd -> Backend -> IO ()
+        readBlockHashOrDefault Best block >>=
+          getCryptographicParametersV2 >>=
+            printResponseValueAsJSON
   where
-    printSuccess (Left x)  = liftIO . putStrLn $ x
-    printSuccess (Right x) = liftIO $ if x then putStrLn "OK" else putStrLn "FAIL"
-    printSuccess' = printSuccess . fmap grpcResponseVal
-
--- |Look up block infos all the way to genesis.
-loop :: Either String Value -> ClientMonad IO ()
-loop v =
-  case v of
-    Left err       -> liftIO $ putStrLn err
-    Right (AE.Object m) ->
-      case KM.lookup "blockParent" m of
-        Just (String parent) -> do
-          printJSON v
-          case KM.lookup "blockSlot" m of
-            Just (AE.Number x) | x > 0 ->
-              fmap grpcResponseVal <$> getBlockInfo parent >>= loop
-            _ -> return () -- Genesis block reached.
-        _ -> error "Unexpected return value for block parent."
-    _ -> error "Unexptected return value for getBlockInfo."
-
+    -- |Print the response value under the provided mapping,
+    -- or fail with an error message if the response contained
+    -- an error.
+    printResponseValue :: (MonadIO m, Show b)
+      => (a -> b)
+      -> GRPCResultV2 (Either String a)
+      -> m ()
+    printResponseValue f res =
+      extractResponseValueOrDie f res >>= liftIO . print
+
+    -- |Print the response value as JSON, or fail with an error
+    -- message if the response contained an error.
+    printResponseValueAsJSON :: (MonadIO m, ToJSON a)
+      => GRPCResultV2 (Either String a)
+      -> m ()
+    printResponseValueAsJSON res = do
+      v <- getResponseValueOrDie res
+      printJSONValues . toJSON $ v
+
+    -- |Print result of a query with side-effects.
+    printSuccess (StatusOk _) = liftIO $ logSuccess ["Success"]
+    printSuccess (StatusNotOk (c, x)) = liftIO $ logError [[i|Non-"OK" status code '#{c}' in response: #{x}|]]
+    printSuccess StatusInvalid = liftIO $ logError [[i|Invalid status code in response|]]
+    printSuccess (RequestFailed x) = liftIO $ logError [[i|Request failed: #{x}|]]
+
+    -- |Parse an account address.
+    parseAccountAddress :: (MonadIO m) => Text -> m ID.AccountAddress
+    parseAccountAddress accAddr =
+      case ID.addressFromText accAddr of
+          Left _ -> logFatal ["Unable to parse account address."]
+          Right a -> return a
+
+    -- |Print info about a block and possibly its ancestors.
+    -- The boolean indicates whether to recurse on the ancestor
+    -- of the block. The recursion bottoms out when the genesis
+    -- block is reached.
+    printBlockInfos :: Bool -> BlockHashInput -> ClientMonad IO ()
+    printBlockInfos recurse bh = do
+      bi <- getResponseValueOrDie =<< getBlockInfoV2 bh
+      printJSONValues $ toJSON bi
+      -- Recurse if were instructed to and if we are not at the genesis block.
+      when (recurse && Queries.biBlockHeight bi /= 0)
+        (printBlockInfos recurse $ Given (Queries.biBlockParent bi))
```

Comment:
Consider moving these `print*` functions to `Output.hs`.

Reply:
I could do that, but I am not entirely sure under which heading they would belong (cfr. the organization of that file).